monitoring/nethealth.go

```diff
@@ -0,0 +1,411 @@
+func (c *nethealthChecker) getNethealthAddr(pods *v1.PodList) (string, error) {
+	// Find nethealth pod with matching host ip address.
+	for _, pod := range pods.Items {
+		if pod.GetLabels()[nethealthLabel] != nethealthValue {
+			continue
+		}
+		if pod.Status.HostIP == c.AdvertiseIP {
+			return fmt.Sprintf("http://%s:%d", pod.Status.PodIP, c.NethealthPort), nil
+		}
+	}
+	return "", trace.NotFound("unable to find local nethealth pod")
+}
+
+// fetchMetrics collects the network metrics from the nethealth pod.
+func fetchMetrics(ctx context.Context, addr string) (metricFamilies map[string]*dto.MetricFamily, err error) {
+	client, err := roundtrip.NewClient(addr, "")
+	if err != nil {
+		return nil, trace.Wrap(err, "failed to connect to nethealth service at %s.", addr)
+	}
+
+	resp, err := client.Get(ctx, client.Endpoint("metrics"), url.Values{})
+	if err != nil {
+		return nil, trace.ConvertSystemError(err)
+	}
+
+	var parser expfmt.TextParser
+	metricsFamilies, err := parser.TextToMetricFamilies(resp.Reader())
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return metricsFamilies, nil
+}
+
+// updateStats updates netStats with new data collected in the provided
+// MetricFamily.
+func (c *nethealthChecker) updateStats(echoRequests, echoTimeouts map[string]float64) (updated []string, err error) {
+	if len(echoRequests) != len(echoTimeouts) {
+		return updated, trace.BadParameter("recieved %d timeout counters and %d request counters",
+			len(echoRequests), len(echoTimeouts))
+	}
+
+	for peer, requests := range echoRequests {
+		timeouts, ok := echoTimeouts[peer]
+		if !ok {
+			return updated, trace.BadParameter("echo timeout data not available for %s", peer)
+		}
+
+		data, err := c.peerStats.Get(peer)
+		if err != nil {
+			return updated, trace.Wrap(err)
+		}
+
+		// Calcuate counter increase since last check and replace previous total
+		requestInc := requests - data.prevRequestTotal
+		data.prevRequestTotal = requests
+		timeoutInc := timeouts - data.prevTimeoutTotal
+		data.prevTimeoutTotal = timeouts
+
+		// Request counter should be constantly increasing, if this is not the
+		// case, nethealth pod most likely reset the counter. Do no update
+		// packet loss percentages this round.
+		if requestInc <= 0 {
+			log.Warn("Request counter did not increase. Nethealth pod may have restarted.")
+			continue
+		}
+
+		// It should not be possible for the timeout counter to have increased
+		// more than the request counter. Log and ignore this situation.
+		if timeoutInc > requestInc {
+			log.WithField("request inc", requestInc).
+				WithField("timeout inc", timeoutInc).
+				Warn("Timeout counter increased more than request counter.")
+			continue
+		}
+
+		// Record the new packet loss percentage, but only keep the last `seriesCapacity` number of data points.
+		if len(data.packetLossPercentages) >= c.seriesCapacity {
+			data.packetLossPercentages = data.packetLossPercentages[1:]
+		}
+		packetLoss := timeoutInc / requestInc
```

Comment:
Interesting, so you're recording the packet loss after each interval and then compare each of them with the threshold below. Do you know how many datapoints nethealth produces b/w the checker runs? I'm just thinking, if it checks it once every 30 seconds, we may have not enough "increases" here and the packet loss will always be either close to 0 or to a 100. Have you considered to instead just accumulating these increases and then calculating packet loss once in "isHealthy" function - so we'd get the packet loss for the entire interval.

Reply:
After thinking about this some more, I think calculating the overall packet loss once for the interval may give some undesired results.

Example:
Two nodes are unable to communicate on the overlay network for some time and end up with 100% packet loss for the entire collection interval. When the overlay network issue is solved and the two nodes are able to communicate, the checker will continue to be fail until the overall packet loss percentage has decreased below the threshold. If starting from 100% this will probably require no packet loss for the entire collection interval.